src/main/java/com/jozufozu/flywheel/backend/engine/uniform/FrameUniforms.java

```diff
@@ -56,39 +88,109 @@ public void write(long ptr) {
 
 		ptr += 96;
 
+		// manage last values of matrices
+		if (!lastInit) {
+			viewPrev.set(view);
+			projectionPrev.set(projection);
+			viewProjectionPrev.set(viewProjectionPrev);
+			cleanProjectionPrev.set(cleanProjection);
+			cleanViewProjectionPrev.set(cleanViewProjection);
+		}
 		ptr = writeMatrices(ptr);
-
-		ptr = writeCamera(ptr, camX, camY, camZ, camera);
+		viewPrev.set(view);
+		projectionPrev.set(projection);
+		viewProjectionPrev.set(viewProjection);
+		cleanProjectionPrev.set(cleanProjection);
+		cleanViewProjectionPrev.set(cleanViewProjection);
+
+		// last values for camera
+		if (!lastInit) {
+			cameraPositionPrev.set(camX, camY, camZ);
+			cameraLookPrev.set(camera.getLookVector());
+			cameraRotPrev.set(camera.getXRot(), camera.getYRot());
+		}
+		ptr = writeCamera(ptr, camX, camY, camZ);
+		cameraPositionPrev.set(camX, camY, camZ);
+		cameraLookPrev.set(camera.getLookVector());
+		cameraRotPrev.set(camera.getXRot(), camera.getYRot());
 
 		var window = Minecraft.getInstance()
 				.getWindow();
-		ptr = writeVec2(ptr, window.getWidth(), window.getHeight());
+		ptr = Uniforms.writeVec2(ptr, window.getWidth(), window.getHeight());
 
 		// default line width: net.minecraft.client.renderer.RenderStateShard.LineStateShard
 		MemoryUtil.memPutFloat(ptr, Math.max(2.5F, (float) window.getWidth() / 1920.0F * 2.5F));
 		ptr += 4;
 
+		MemoryUtil.memPutFloat(ptr, (float) window.getWidth() / (float) window.getHeight());
+		ptr += 4;
+
+		MemoryUtil.memPutFloat(ptr, Minecraft.getInstance().gameRenderer.getDepthFar());
+		ptr += 4;
+
 		MemoryUtil.memPutInt(ptr, getConstantAmbientLightFlag(context));
```

Comment:
Constant ambient light should be moved to level uniforms.

Reply:
Good point. It's moved now.